modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java

```diff
@@ -17,74 +17,162 @@
+import static org.apache.ignite.spi.communication.tcp.TcpCommunicationSpi.SENT_MESSAGES_METRIC_DESC;
+import static org.apache.ignite.spi.communication.tcp.TcpCommunicationSpi.SENT_MESSAGES_METRIC_NAME;
+import static org.apache.ignite.spi.communication.tcp.TcpCommunicationSpi.receivedMessagesByTypeMetricName;
+import static org.apache.ignite.spi.communication.tcp.TcpCommunicationSpi.sentMessagesByTypeMetricName;
+
 /**
  * Statistics for {@link org.apache.ignite.spi.communication.tcp.TcpCommunicationSpi}.
  */
-public class TcpCommunicationMetricsListener implements GridNioMetricsListener{
-    /** Counter factory. */
-    private static final Callable<LongHolder> HOLDER_FACTORY = new Callable<LongHolder>() {
-        @Override public LongHolder call() {
-            return new LongHolder();
-        }
-    };
+class TcpCommunicationMetricsListener {
+    /** Metrics manager. */
+    private final GridMetricManager mmgr;
 
-    /** Received bytes count. */
-    private final LongAdder rcvdBytesCnt = new LongAdder();
+    /** Metrics registry. */
+    private final MetricRegistry mreg;
 
-    /** Sent bytes count.*/
-    private final LongAdder sentBytesCnt = new LongAdder();
+    /** */
+    private final Function<Short, AtomicLongMetric> sentMsgsCntByTypeMetricFactory;
 
-    /** All registered metrics. */
-    private final Set<ThreadMetrics> allMetrics = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    /** */
+    private final Function<Short, AtomicLongMetric> rcvdMsgsCntByTypeMetricFactory;
 
-    /** Thread-local metrics. */
-    private final ThreadLocal<ThreadMetrics> threadMetrics = new ThreadLocal<ThreadMetrics>() {
-        @Override protected ThreadMetrics initialValue() {
-            ThreadMetrics metrics = new ThreadMetrics();
+    /** */
+    private final Function<Object, AtomicLongMetric> sentMsgsCntByConsistentIdMetricFactory;
 
-            allMetrics.add(metrics);
+    /** */
+    private final Function<Object, AtomicLongMetric> rcvdMsgsCntByConsistentIdMetricFactory;
 
-            return metrics;
-        }
-    };
+    /** Sent bytes count metric.*/
+    private final AtomicLongMetric sentBytesMetric;
+
+    /** Received bytes count metric. */
+    private final AtomicLongMetric rcvdBytesMetric;
+
+    /** Sent messages count metric. */
+    private final AtomicLongMetric sentMsgsMetric;
+
+    /** Received messages count metric. */
+    private final AtomicLongMetric rcvdMsgsMetric;
+
+    /** Sent messages count metrics grouped by message type. */
+    ConcurrentHashMap<Short, AtomicLongMetric> sentMsgsMetricsByType = new ConcurrentHashMap<>();
+
+    /** Received messages count metrics grouped by message type. */
+    ConcurrentHashMap<Short, AtomicLongMetric> rcvdMsgsMetricsByType = new ConcurrentHashMap<>();
+
+    /**
+     * Sent messages count metrics grouped by message node id.
+     * @deprecated for JMX support only.
```

Comment:
Let's start sentence with the capital letter and explain to the reader, what he(or she) should use instead of this map.